int16[] hole_center_x
int16[] hole_center_y
int32[] point_index
int16[] circle_color
